unittests/ExecutionEngine/JitWrapper.cpp

```diff
@@ -0,0 +1,175 @@
+
+#include "gc/ExecutionEngine/JitWrapper/Module.hpp"
+#include "mlir/AsmParser/AsmParser.h"
+#include "mlir/ExecutionEngine/MemRefUtils.h"
+#include "mlir/IR/AsmState.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Pass/PassManager.h"
+#include "llvm/Support/ErrorOr.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/SourceMgr.h"
+#include "llvm/Support/raw_ostream.h"
+#include "gtest/gtest.h"
+#include <memory>
+
+using namespace mlir;
+
+static const char code1[] = R"mlir(
+module {
+llvm.mlir.global constant @__num_orig_num_args(3 : i32) : i32
+func.func @compute(%a: tensor<128xf32>, %b: tensor<128xf32>) -> tensor<128xf32> attributes { llvm.emit_c_interface } {
+    %out = tensor.empty() : tensor<128xf32>
+    %2 = linalg.add ins(%a, %b : tensor<128xf32>,tensor<128xf32>) outs(%out : tensor<128xf32>) -> tensor<128xf32>
+    return %2 : tensor<128xf32>
+}
+}
+)mlir";
+
+extern "C" {
+extern int gc_runtime_keep_alive;
+}
+
+TEST(ExecutionEngine, JitWrapper) {
+  gc_runtime_keep_alive = 0;
+  MLIRContext ctx{gc::initAndGetDialects()};
+  std::unique_ptr<llvm::MemoryBuffer> ir_buffer =
+      llvm::MemoryBuffer::getMemBuffer(code1);
+  // Parse the input mlir.
+  llvm::SourceMgr sourceMgr;
+  sourceMgr.AddNewSourceBuffer(std::move(ir_buffer), llvm::SMLoc());
+  mlir::OwningOpRef<mlir::ModuleOp> module =
+      mlir::parseSourceFile<mlir::ModuleOp>(sourceMgr, &ctx);
+  ASSERT_TRUE(module);
+  auto jited = gc::JitModule::create(module.get());
+  bool jit_success = static_cast<bool>(jited);
+  if (!jit_success) {
+    auto err = jited.takeError();
+    llvm::errs() << err;
+    llvm::consumeError(std::move(err));
+  }
+  ASSERT_TRUE(jit_success);
+  OwningMemRef<float, 1> bufA{
+      {128}, {128}, [](float &ptr, ArrayRef<int64_t>) { ptr = 1.0f; }};
+  OwningMemRef<float, 1> bufB{
+      {128}, {128}, [](float &ptr, ArrayRef<int64_t> idx) { ptr = idx[0]; }};
+  OwningMemRef<float, 1> bufC{{128}, {128}};
+  void *args[] = {&*bufA, &*bufB, &*bufC};
+  jited.get()->call(args);
+  for (int i = 0; i < 128; i++) {
+    ASSERT_EQ(bufC[{i}], 1.0f + i);
+  }
+}
+
+// compute d = (a+a) + (b+b) + c, where a,b is marked constant
+// bufIdx: a=0, b=1, c=2, d=3, foldedA=4, foldedB=5
+static const char code2[] = R"mlir(
+module {
+llvm.mlir.global constant @__num_orig_num_args(4 : i32) : i32
+llvm.mlir.global constant @__fold_buffer_ids(dense<[2, 114514, 1919810]> : tensor<3 x i64>) : !llvm.array<3 x i64>
+// a,b, foldedA,foldedB
+llvm.mlir.global constant @__fold_args(dense<[4, 0, 1, 4, 5]> : tensor<5xi32>) : !llvm.array<5 x i32>
+// foldedA, foldedB, c, d
+llvm.mlir.global constant @__compute_args(dense<[4, 4, 5, 2, 3]> : tensor<5xi32>) : !llvm.array<5 x i32>
+
+func.func @fold(%a: tensor<128xf32>, %b: tensor<128xf32>) -> (tensor<128xf32>, tensor<128xf32>) attributes { llvm.emit_c_interface } {
+    %c0 = arith.constant 0 : index
+    cpuruntime.printf "HI%zu\n" %c0 : index
```

Comment:
We print "HI0" in `fold()`, and we check if there is "HI0" in stdout for 1st run, 2nd run, and 3rd run (after evicting cache)

Reply:
Because `%c0 = arith.constant 0 : index` and `cpuruntime.printf "HI%zu\n" %c0 : index` are not related to any constant input, I think they should **not** be processed by pass `constant-weights-folding`.

Reply:
This is the IR **after** `constant-weights-folding`. I used `printf` to find if the `fold()` code is executed or not.

Reply:
The problem is that whether a constant operation which is not related to any constant input should be moved to `fold`. Currently I leave it in `compute`.

Reply:
Here the code is the manual written IR for testing, not the real output of  `constant-weights-folding`. If I use the real output `constant-weights-folding`, how can I test if the `fold()` func is skipped at the second execution?

Reply:
OK I got you. Let's use this manually written example as a unit test.